[하지현] Assignment1: Express API - 초기 환경 세팅

haaazzi opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 데이터베이스 작업
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표
- Express 초기 환경 세팅

<br />

## :: 구현 사항 설명 
1) Express 초기 환경 설정하기
 `Express` 설치 / 적용
 `nodemon` 설치 / 적용
 `cors` 설치 / 적용
 `dotenv` 설치 / 적용
 `morgan` 설치 / 적용
2) TypeORM, dbConnection 설정
 `TypeORM` 설치 및 적용
 `dbConnection` 정상 작동 확인

<br />

## :: 테스트 결과 이미지
1. 터미널 결과
![스크린샷 2023-04-17 오후 10 32 10](https://user-images.githubusercontent.com/127677568/232502306-fdda2847-f230-4e3c-81a7-f7ff23479937.png)
2. Postman 테스트 결과
![스크린샷 2023-04-17 오후 10 44 03](https://user-images.githubusercontent.com/127677568/232502479-5f27c257-7f6e-4eb2-9d0d-c2e01139eb40.png)

<br />

Commit:
ADD: initial-setting

kyeongraekim commented on jihyeon/.env.sample:
```diff
@@ -0,0 +1,9 @@
+PORT = YOUR PORT NUMBER SHOULD BE HERE!
+
+DB_HOST= host #
+DB_USER=root
+DB_PASS=password
+DB_CONNECTION = mysql
+DB_DATABASE = database
+DB_PORT = 3306
+DB_LOGGING = TRUE
```
- 전체적으로 = 주변에 Whitespace가 균일하지 못하게 설정되어 있네요!
- 일괄적으로 환경변수내 키-밸류 값내에서의 whitespace를 모두 삭제해 주시는 것을 권장드립니다!

kyeongraekim commented on jihyeon/app.js:
```diff
@@ -0,0 +1,40 @@
+const express = require('express');
+const cors = require('cors');
+const morgan = require ("morgan");
+const http = require ("http");
+require("dotenv").config();
+
```
```suggestion
require('dotenv').config();

const logger = require('morgan');
const express = require('express');
const cors = require('cors');
const { DataSource } = require('typeorm');

```

- 위 내용과 같이 환경변수를 호출하는 config 문을 제일 위에서 선언해주세요!
- 코드 시작과 동시에 필요한 환경변수 세팅을 확실히 하기 위한 컨벤션이라고 생각해주시면 되겠습니다!

Jon9Beom commented on jihyeon/.env.sample:
```diff
@@ -0,0 +1,9 @@
+PORT = YOUR PORT NUMBER SHOULD BE HERE!
+
+DB_HOST= host #
+DB_USER=root
+DB_PASS=password
+DB_CONNECTION = mysql
+DB_DATABASE = database
+DB_PORT = 3306
+DB_LOGGING = TRUE
```
```suggestion
PORT = YOUR PORT NUMBER SHOULD BE HERE!

DB_HOST= host #
DB_USER=db_user
DB_PASSWORD=password
DB_CONNECTION = rdbms_type
DB_DATABASE = database
DB_PORT = db_port
DB_LOGGING = true / false
```
1. root 와같은 database user name을 중의적으로 표현해주세요.
2. DB_PASSWORD라고 명확히 명칭해주세요!
3. 어떤 rdbms를 쓰는지 해당 rdbms의 port number는 무엇인지 모두 직접 명시하지 않고 중의적으로 표현해주세요.

Jon9Beom replied:
dbmate의 타겟 주소 목적의 `DATABASE_URL` 변수 또한 중의적인 표현을 활용하여 작성해주세요!
```jsx
// Example	

DATABASE_URL="rdbmsType://username:password@iphost:portNumber/databaseName"
```

Jon9Beom commented on jihyeon/app.js:
```diff
@@ -0,0 +1,40 @@
+const express = require('express');
+const cors = require('cors');
+const morgan = require ("morgan");
+const http = require ("http");
+require("dotenv").config();
```
```suggestion
require("dotenv").config();

const express = require('express');
const cors = require('cors');
const morgan = require ("morgan");
```
1. 서버가 실행될 때, 초기에 환경변수를 읽어오는 코드는 보통 상단에 작성됩니다.
2. http 모듈은 node.js의 built-in 모듈로서, 인터넷을 통해 들어오는 네트워크 요청을 코드로서 처리할 수 있도록 할 수 있게 해주는 모듈입니다. 우리는 http의 기능에 더불어 보다 더 많은 기능을, 빠르고 쉽게 제공해주는 Express를 활용하기 때문에 http 모듈은 현재 필요하지 않습니다.

Jon9Beom commented on jihyeon/app.js:
```diff
@@ -0,0 +1,40 @@
+    database: process.env.DB_DATABASE
+})
+
+dataSource.initialize()
+    .then(() => {
+       console.log("Data Source has been initialized!")
+   })
```
- 추가로 `dataSource.initialize()` 함수가 실행되고 나면, then()과 catch()를 사용해서 에러를 핸들링해줄 수 있습니다. 이렇게 하면, initialize() 함수가 실행되어서 database와 실제로 connection을 생성할 때, 성공했는지 실패했는지 여부를 확인할 수 있습니다.
- typeorm 공식문서의 initialize() 함수의 사용 부분을 참고하여 코드를 수정해주세요!

Jon9Beom commented on jihyeon/app.js:
```diff
@@ -0,0 +1,40 @@
+app.use(cors());
+app.use(morgan('dev'));
+
+app.get('/ping', function (req, res, next){
+    res.json({message: 'pong'})
+});
```
```suggestion
app.get('/ping', function (req, res, next){
    return res.status(200).json({message: 'pong'})
});
```
http request method에 적합한 status code 또한 백엔드에서 API 별로 지정해줄 수 있습니다.

Jon9Beom commented on jihyeon/app.js:
```diff
@@ -0,0 +1,40 @@
+    res.json({message: 'pong'})
+});
+
+const server = http.createServer(app);
```
```suggestion
```
http는 쓰지 않도록 해주세요~

Jon9Beom commented on jihyeon/app.js:
```diff
@@ -0,0 +1,40 @@
+const PORT = process.env.PORT;
+
+const start = async () => {
+    server.listen(PORT, () => console.log(`server is listening on ${PORT}`));
```
```suggestion
    app.listen(PORT, () => console.log(`server is listening on ${PORT}`));
```
http를 쓰지 않기 때문에, express로 만들어준 app이라는 변수를 활용하여 listen하면 됩니다.

Commit:
MODIFY: .env.sample / app.js 수정
1) .env.sample
  - whitespace 제거
  - 중의적인 표현으로 변경
2) app.js
  - http 관련 필요없는 코드 제거
  - status code 추가
  - 에러 핸들링 : catch() 사용

Jon9Beom commented:
LGTM 👍
